src/query_graph/graph_path.rs

```diff
@@ -430,6 +430,25 @@ impl Display for OpGraphPathContext {
 #[derive(Clone)]
 pub(crate) struct SimultaneousPaths(pub(crate) Vec<Arc<OpGraphPath>>);
 
+impl SimultaneousPaths {
+    pub(crate) fn fmt_indented(&self, f: &mut Formatter<'_>, indent: &str) -> std::fmt::Result {
```

Comment:
Is this only for potential debug logging? If not and it is in a hot path: it’s possible to refactor this to not allocate intermediate strings by writing components to `&mut Formatter` one by one. `src/query_plan/display.rs` is an example of this pattern. (`current_indent_level` there enables arbitrarily-nested indentation, but is not needed if there’s exactly one ident level.)

Reply:
It's used for debug logging. However, logging at runtime is still a hot path. Let me improve it a bit.

BTW, the indentation is not used for now, but it was used somewhere else in the original TypeScript implementation. So, I tried to port the indentation feature just in case.